File: src/main/java/top/chiloven/lukosbot2/platform/telegram/TelegramSender.java
package top.chiloven.lukosbot2.platform.telegram;

import org.telegram.telegrambots.meta.api.methods.send.SendDocument;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.methods.send.SendPhoto;
import org.telegram.telegrambots.meta.api.objects.InputFile;
import top.chiloven.lukosbot2.model.message.media.BytesRef;
import top.chiloven.lukosbot2.model.message.media.MediaRef;
import top.chiloven.lukosbot2.model.message.media.PlatformFileRef;
import top.chiloven.lukosbot2.model.message.media.UrlRef;
import top.chiloven.lukosbot2.model.message.outbound.*;
import top.chiloven.lukosbot2.platform.ISender;
import top.chiloven.lukosbot2.util.message.OutboundPartUtils;

import java.io.ByteArrayInputStream;
import java.util.Collections;
import java.util.List;

/**
 * Telegram sender that translates {@link OutboundMessage} into Telegram API calls.
 *
 * <p>Supports ordered, mixed content via {@link OutboundMessage#parts()}:
 * text, image and file. When {@link DeliveryHints#preferCaption()} is enabled, the sender will try to merge an adjacent
 * text part into a media caption for a better user experience.</p>
 */
public final class TelegramSender implements ISender {

    private final TelegramStack stack;

    public TelegramSender(TelegramStack stack) {
        this.stack = stack;
    }

    @Override
    public void send(OutboundMessage out) {
        if (out == null) return;

        String chatId = String.valueOf(out.addr().chatId());
        List<OutPart> parts = out.parts() == null ? Collections.emptyList() : out.parts();
        if (parts.isEmpty()) return;

        List<OutPart> normalized = OutboundPartUtils.mergeAdjacentTextParts(parts);

        boolean preferCaption = out.hints() != null && out.hints().preferCaption();

        for (int i = 0; i < normalized.size(); i++) {
            OutPart p = normalized.get(i);
            switch (p) {
                case OutText(String text1) -> {
                    String text = OutboundPartUtils.safeText(text1);
                    if (text.isBlank()) continue;

                    // If next part is media and prefers caption, try to attach this text as caption.
                    if (preferCaption && i + 1 < normalized.size()) {
                        OutPart next = normalized.get(i + 1);
                        if (next instanceof OutImage img && (img.caption() == null || img.caption().isBlank())) {
                            sendPhoto(chatId, img, text);
                            i++; // consume next
                            continue;
                        }
                        if (next instanceof OutFile f && (f.caption() == null || f.caption().isBlank())) {
                            sendDocument(chatId, f, text);
                            i++; // consume next
                            continue;
                        }
                    }

                    sendText(chatId, text);
                }
                case OutImage img -> sendPhoto(chatId, img, OutboundPartUtils.safeText(img.caption()));
                case OutFile f -> sendDocument(chatId, f, OutboundPartUtils.safeText(f.caption()));
                case null, default -> {
                }
            }

        }
    }

    private void sendPhoto(String chatId, OutImage img, String caption) {
        if (img == null) return;
        SendPhoto sp = SendPhoto.builder()
                .chatId(chatId)
                .photo(toInputFile(img.ref(), img.name(), img.mime()))
                .caption(caption == null || caption.isBlank() ? null : caption)
                .build();
        stack.execute(sp);
    }

    private void sendDocument(String chatId, OutFile f, String caption) {
        if (f == null) return;
        SendDocument sd = SendDocument.builder()
                .chatId(chatId)
                .document(toInputFile(f.ref(), f.name(), f.mime()))
                .caption(caption == null || caption.isBlank() ? null : caption)
                .build();
        stack.execute(sd);
    }

    private void sendText(String chatId, String text) {
        if (text == null || text.isBlank()) return;
        SendMessage sm = SendMessage.builder()
                .chatId(chatId)
                .text(text)
                .build();
        stack.execute(sm);
    }

    private static InputFile toInputFile(MediaRef ref, String name, String mime) {
        switch (ref) {
            case null -> {
                return new InputFile("about:blank");
            }
            case BytesRef b -> {
                String n = OutboundPartUtils.pickMediaName(name, b.name(), mime, true);
                return new InputFile(new ByteArrayInputStream(b.bytes()), n);
            }
            case UrlRef(String url) -> {
                return new InputFile(url);
            }
            case PlatformFileRef p -> {
                // Telegram accepts file_id in the same field.
                return new InputFile(p.fileId());
            }
            default -> {
            }
        }

        return new InputFile("about:blank");
    }

}

File: platform/telegram/src/main/java/top/chiloven/lukosbot2/platform/telegram/TelegramSender.kt
/*
 * Copyright © 2026 Chiloven945
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package top.chiloven.lukosbot2.platform.telegram

import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.telegram.telegrambots.meta.api.methods.send.SendDocument
import org.telegram.telegrambots.meta.api.methods.send.SendMessage
import org.telegram.telegrambots.meta.api.methods.send.SendPhoto
import org.telegram.telegrambots.meta.api.objects.InputFile
import top.chiloven.lukosbot2.core.model.message.media.BytesRef
import top.chiloven.lukosbot2.core.model.message.media.MediaRef
import top.chiloven.lukosbot2.core.model.message.media.PlatformFileRef
import top.chiloven.lukosbot2.core.model.message.media.UrlRef
import top.chiloven.lukosbot2.core.model.message.outbound.OutFile
import top.chiloven.lukosbot2.core.model.message.outbound.OutImage
import top.chiloven.lukosbot2.core.model.message.outbound.OutText
import top.chiloven.lukosbot2.core.model.message.outbound.OutboundMessage
import top.chiloven.lukosbot2.platform.ISender
import top.chiloven.lukosbot2.util.message.OutboundPartUtils
import java.io.ByteArrayInputStream

/**
 * Telegram sender that translates [OutboundMessage] into Telegram API calls.
 *
 * <p>Supports ordered, mixed content via {@link OutboundMessage#parts()}:
 * text, image and file. When {@code DeliveryHints.preferCaption()} is enabled, the sender will try to merge an adjacent
 * text part into a media caption for a better user experience.</p>
 */
internal class TelegramSender(
    private val stack: TelegramStack,
    private val blockingDispatcher: CoroutineDispatcher = Dispatchers.IO,
) : ISender {

    override suspend fun send(out: OutboundMessage) {
        withContext(blockingDispatcher) {
            val chatId = out.addr().chatId().toString()
            val parts = out.parts().orEmpty()
            if (parts.isEmpty()) return@withContext

            val normalized = OutboundPartUtils.mergeAdjacentTextParts(parts)

            val preferCaption = out.hints() != null && out.hints().preferCaption()

            var i = 0
            while (i < normalized.size) {
                when (val p = normalized[i]) {
                    is OutText -> {
                        val text = OutboundPartUtils.safeText(p.text())
                        if (text.isBlank()) {
                            i++
                            continue
                        }

                        // If next part is media and prefers caption, try to attach this text as caption.
                        if (preferCaption && i + 1 < normalized.size) {
                            val next = normalized[i + 1]
                            if (next is OutImage && (next.caption().isNullOrBlank())) {
                                sendPhoto(chatId, next, text)
                                i += 2 // consume next
                                continue
                            }

                            if (next is OutFile && next.caption().isNullOrBlank()) {
                                sendDocument(chatId, next, text)
                                i += 2 // consume next
                                continue
                            }
                        }

                        sendText(chatId, text)
                    }

                    is OutImage -> sendPhoto(
                        chatId,
                        p,
                        OutboundPartUtils.safeText(p.caption())
                    )
                    is OutFile -> sendDocument(
                        chatId,
                        p,
                        OutboundPartUtils.safeText(p.caption())
                    )
                }
                i++
            }
        }
    }

    private fun sendPhoto(
        chatId: String,
        img: OutImage?,
        caption: String
    ) {
        if (img == null) return
        val sp = SendPhoto.builder()
                .chatId(chatId)
                .photo(toInputFile(img.ref(), img.name(), img.mime()))
                .caption(caption.ifBlank { null })
                .build()
        stack.execute(sp)
    }

    private fun sendDocument(
        chatId: String,
        f: OutFile?,
        caption: String
    ) {
        if (f == null) return
        val sd = SendDocument.builder()
                .chatId(chatId)
                .document(toInputFile(f.ref(), f.name(), f.mime()))
                .caption(caption.ifBlank { null })
                .build()
        stack.execute(sd)
    }

    private fun sendText(chatId: String, text: String) {
        if (text.isBlank()) return
        val sm = SendMessage.builder()
                .chatId(chatId)
                .text(text)
                .build()
        stack.execute(sm)
    }

    private fun toInputFile(
        ref: MediaRef?,
        name: String?,
        mime: String?
    ): InputFile {
        return when (ref) {
            null -> InputFile("about:blank")
            is BytesRef -> {
                val n = OutboundPartUtils.pickMediaName(
                    name,
                    ref.name(),
                    mime,
                    true
                )
                InputFile(ByteArrayInputStream(ref.bytes()), n)
            }

            is UrlRef -> InputFile(ref.url())
            is PlatformFileRef -> {
                // Telegram accepts file_id in the same field.
                InputFile(ref.fileId())
            }
        }
    }

}
